### DISCUSSION TITLE
Rate Group Slipping

### DISCUSSION THREAD

**Discussion by noahrose18** [nasa/fprime#2340] [Q&A] [answered]
Hello!

I am trying to create a rate group that supports various interrupt frequencies (1/4Hz, 1/2Hz, 1Hz, 10Hz, and 25Hz). I thus set the clock of the overall system to the Lowest common multiple (50Hz), and I created an event within the 10Hz handler to validate that it is working correctly.

I am noticing that the timing of the 10Hz event (as seen within the fprime-gds popup events tab) is slipping, aka it is interrupting slighlty late every iteration. I am wondering if this is the expected behavior and, if not, how to make these interrupts hit at the same exact interval every time.

**Comment by timcanham** [comment]
@noahrose18 Just some clarification: The terminology would be rate _groups_, at 1/4Hz, 1/2Hz, etc. The typical approach would be to drive the `Svc/RateGroupDriver` with your base 50Hz and then use the dividers to call your `Svc/ActiveRateGroup` instances at the rates you want. Are you doing this?

**Reply by timcanham** [reply]
In your example, you would have `Svc/ActiveRateGroup` component instance for each of your rate groups, i.e, 1/4Hz, 1/2Hz, etc.

**Reply by noahrose18** [reply]
Thanks for the quick response Tim! Yes I am doing this. More specifically:

I am driving the SVC/RateGroupDriver at 50Hz with startSimulatedCycle(20); // 20 ms cycle = 50Hz

The dividers then are 200, 100, 50, 5, and 2 respectively.

**Reply by timcanham** [reply]
If you are using an RTOS with priorities, you would set the task priority for the ActiveRateGroup instance based on how frequently each rate group has to run. The highest priority would be assigned to 25Hz, the next to 10Hz, etc. If that's what you have done, then the 50Hz would always run before the 25Hz and make it slightly delayed each time. If it's Linux and you are in user space, the scheduler may just be round-robining the rate groups, so sometimes your 10Hz will be later than you expected. What OS are you using?

**Reply by noahrose18** [reply]
I'm using Linux. I'm seeing that when I ask for a ISR at a frequency of 10Hz, it is actually executing slightly less than 10Hz (say 9.5Hz). So, this is slightly different than the error that you are describing, right?

**Reply by timcanham** [reply]
Yes, although what I'm describing isn't an error in an RTOS, it's how priority scheduling of tasks works. What is your "ISR" in this context?

**Reply by noahrose18** [reply]
The ISR is the Handler executing at 10Hz. It logs one event and sends out data to two ports.

**Reply by timcanham** [reply, marked as answer]
Is the ISR driven by a hardware timer? Something else?

If you are doing the pattern:

```c
while (1) {
   doSomething();
   sleep(100ms);
}
```
It will slow the loop since you are executing the logic then sleeping.

**Reply by noahrose18** [reply]
Looks like this is what is happening in the main.cpp code. Thanks!

**Reply by timcanham** [reply]
There is a `Svc/LinuxTimer` implementation in the core repo based on the Linux `timerfd` feature that you can use to get repeatable cycles.

**Reply by timcanham** [reply]
https://github.com/nasa/fprime/tree/devel/Svc/LinuxTimer
